### PR TITLE
[SYCL][XPTI] Add unique queue ID to trace

### PR DESCRIPTION
**Pull request by KseniyaTikhomirova**

### CODE REVIEW

File: sycl/source/detail/queue_impl.hpp [hunk 1]
```diff
@@ -196,6 +199,10 @@ class queue_impl {
       // This section is the second part of the instrumentation that uses the
       // tracepoint information and notifies
     }
+// #if XPTI_ENABLE_INSTRUMENTATION
+// if (!MHostQueue)
+//     xpti::addMetadata(PrepareNotify.traceEvent(), "handle", getHandleRef());
```
Comment (KseniyaTikhomirova) [inline sycl/source/detail/queue_impl.hpp]: queue_handle

File: sycl/source/detail/queue_impl.hpp [hunk 2]
```diff
@@ -255,6 +263,10 @@ class queue_impl {
           make_error_code(errc::invalid),
           "Device provided by native Queue not found in Context.");
     }
+#if XPTI_ENABLE_INSTRUMENTATION
+if (!MHostQueue)
+    xpti::addMetadata(PrepareNotify.traceEvent(), "handle", getHandleRef());
+#endif
```
Comment (KseniyaTikhomirova) [inline sycl/source/detail/queue_impl.hpp:265]: add a proper description to destructor that event and all metadata will be destroyed after

File: sycl/source/CMakeLists.txt [hunk 3]
```diff
@@ -3,6 +3,7 @@
 #2. Use AddLLVM to modify the build and access config options
 #cmake_policy(SET CMP0057 NEW)
 #include(AddLLVM)
+add_definitions(-O0 -g)
```
Comment (KseniyaTikhomirova) [inline sycl/source/CMakeLists.txt:6]: remove

File: sycl/source/detail/queue_impl.hpp [hunk 4]
```diff
@@ -748,6 +757,9 @@ class queue_impl {
                     const std::shared_ptr<queue_impl> &SecondaryQueue,
                     const detail::code_location &Loc,
                     const SubmitPostProcessF *PostProcess) {
+                      std::cout << "PrimaryQueue = " << PrimaryQueue << std::endl;
```
Comment (KseniyaTikhomirova) [inline sycl/source/detail/queue_impl.hpp:760]: remove

File: sycl/source/detail/scheduler/commands.cpp [hunk 5]
```diff
@@ -64,6 +64,17 @@ bool CurrentCodeLocationValid() {
   return (FileName && FileName[0] != '\0') ||
          (FunctionName && FunctionName[0] != '\0');
 }
+
+void emitInstrumentation(uint32_t StreamID, QueueImplPtr Queue, uint64_t InstanceID, xpti_td* TraceEvent, uint16_t Type, const char *Txt) {
+  if (!(xptiCheckTraceEnabled(StreamID, Type) && TraceEvent))
+    return;
+  // Trace event notifier that emits a Type event
+  xpti::addMetadata(TraceEvent, "queue_id",
```
Comment (KseniyaTikhomirova) [inline sycl/source/detail/scheduler/commands.cpp:72]: move to the place where other metadata placed, adding it here trigger duplication path

File: sycl/source/handler.cpp [hunk 6]
```diff
@@ -236,10 +236,19 @@ event handler::finalize() {
       std::vector<sycl::detail::pi::PiEvent> RawEvents;
       detail::EventImplPtr NewEvent;
 
+    #ifdef XPTI_ENABLE_INSTRUMENTATION
+      //uint32_t StreamID, uint64_t InstanceID, xpti_td* TraceEvent,
```
Comment (KseniyaTikhomirova) [inline sycl/source/handler.cpp:240]: consider XPTIScope usage here

File: sycl/unittests/xpti_trace/CMakeLists.txt [hunk 7]
```diff
@@ -1,8 +1,9 @@
 add_subdirectory(xptitest_subscriber)
 
-add_definitions(-DXPTI_ENABLE_INSTRUMENTATION)
+add_definitions(-DXPTI_ENABLE_INSTRUMENTATION -O0 -g)
```
Comment (KseniyaTikhomirova) [inline sycl/unittests/xpti_trace/CMakeLists.txt:3]: remove

File: sycl/unittests/xpti_trace/xptitest_subscriber/XPTISubscriber.cpp [hunk 8]
```diff
@@ -62,7 +66,76 @@ XPTI_CALLBACK_API void testCallback(uint16_t TraceType,
...
+        xpti::metadata_t *Metadata = xptiQueryMetadata(Event);
+        for (const auto &Item : *Metadata) {
+          std::string_view Key{xptiLookupString(Item.first)};
+          if (Key == "queue_id") {
+            Message.append(std::string("task_begin:") + Key.data() + std::string(":") + std::to_string(xpti::getMetadata<unsigned long long>(Item).second));
+          }
+        }
+        GReceivedNotifications.push_back(
+                std::make_pair(TraceType, Message));
+                // "memory_object", "kernel_name"
```
Comment (KseniyaTikhomirova) [inline sycl/unittests/xpti_trace/xptitest_subscriber/XPTISubscriber.cpp:121]: remove

**Comment by KseniyaTikhomirova** [comment]
sycl-trace run example:
`bash-5.1$ ./build/bin/sycl-trace --sycl --print-format=verbose ./a.out 
[SYCL] Queue create: 
        queue_handle: 0x1b11a18
        queue_id: 0
        is_inorder: 1
        sycl_device: 0x1c3fba8
        sycl_device_name: Intel(R) Xeon(R) Platinum 8380 CPU @ 2.30GHz
        sycl_context: 0x1a97728
[SYCL] Queue create: 
        queue_id: 1
        is_inorder: 0
        sycl_device: 0
        sycl_device_name: SYCL host device
        sycl_context: 0
[SYCL] Queue destroy: 
        queue_handle: 0x1b11a18
        queue_id: 0
[SYCL] Queue destroy: 
        queue_id: 1`

File: sycl/source/detail/queue_impl.hpp [hunk 9]
```diff
@@ -255,6 +224,43 @@ class queue_impl {
...
+                            MDevice->getDeviceName());
+          xpti::addMetadata(
+              TEvent, "sycl_device",
+              reinterpret_cast<size_t>(
+                  MDevice->is_host() ? 0 : MDevice->getHandleRef()));
+        }
+        xpti::addMetadata(TEvent, "is_inorder", MIsInorder);
+        xpti::addMetadata(TEvent, "queue_id", MQueueID);
+        if (!MHostQueue)
+          xpti::addMetadata(PrepareNotify.traceEvent(), "queue_handle",
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/queue_impl.hpp]: https://github.com/intel/llvm/pull/11548/commits/8e7693c9a865e4e3ae506432a18ac0fc1eb23930

File: sycl/source/detail/queue_impl.hpp [hunk 10]
```diff
@@ -231,10 +192,18 @@ class queue_impl {
                   MDevice->is_host() ? 0 : MDevice->getHandleRef()));
         }
         xpti::addMetadata(TEvent, "is_inorder", MIsInorder);
+        xpti::addMetadata(TEvent, "queue_id", MQueueID);
+        if (!MHostQueue)
+          xpti::addMetadata(PrepareNotify.traceEvent(), "queue_handle",
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/queue_impl.hpp]: https://github.com/intel/llvm/pull/11548/commits/8e7693c9a865e4e3ae506432a18ac0fc1eb23930

File: sycl/unittests/xpti_trace/QueueIDCheck.cpp [hunk 11]
```diff
@@ -0,0 +1,189 @@
+//==------------ NodeCreation.cpp --- XPTI integration unit tests ----------==//
```
Comment (sergey-semenov) [inline sycl/unittests/xpti_trace/QueueIDCheck.cpp:1]:
```suggestion
//==------------ QueueIDCheck.cpp --- XPTI integration unit tests ----------==//
```
Not blocking

**Review by sergey-semenov — APPROVED** [review]

**Comment by againull** [comment]
@KseniyaTikhomirova Could you please fix post-commit failures after your commit: https://github.com/intel/llvm/actions/runs/6658855529/job/18096640484

**Comment by againull** [comment]
> @KseniyaTikhomirova Could you please fix post-commit failures after your commit: https://github.com/intel/llvm/actions/runs/6658855529/job/18096640484

Fixed here https://github.com/intel/llvm/pull/11682

File: sycl/unittests/xpti_trace/QueueIDCheck.cpp (hunk 11 above)
Comment (KseniyaTikhomirova) [reply sycl/unittests/xpti_trace/QueueIDCheck.cpp:1]: fixing here https://github.com/intel/llvm/pull/11690

**Pull request merged**